## Student: Deng, Zixuan (V00971633)
# Problem 3. Associations [30 points]


### [Retail; 15 points] 
Now, get the online retail dataset from http://archive.ics.uci.edu/
ml/datasets/Online+Retail, and use the Apriori algorithm implemented in apyori package
to mine the top 5 associations for each country in the dataset. What are your findings?

In [1]:
import numpy as np
import pandas as pd


# import data
df = pd.read_excel('./Online Retail.xlsx')

print(df.shape)
df.head()

(541909, 8)


InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

Would like to preprocess the data into new format


---



invoiceNo    | StockCode  |Country

536365    |8231A, 2334, 89098T  |UK

In [2]:
def combine_stock_code_country(df, country):
    """

    return a new dataframe that in the same country, each invoice has combined, has some new stock code
    """
    country_df = df[df.Country == country]


    invoicno_country = country_df.InvoiceNo.value_counts()




    new_df = pd.DataFrame([], columns = ['new_invoice_no', 'new_stock_code'])

    new_invoice = invoicno_country.index
    new_df['new_invoice_no'] = new_invoice
    new_stock_code = []
    for d in invoicno_country.index: 

        stockcode = []

        for i in (country_df[country_df.InvoiceNo == d]).StockCode:
            stockcode.append(str(i))
        new_stock_code.append(stockcode)
        # print(stockcode)
        

    new_df['new_stock_code'] = new_stock_code


    return new_df, new_stock_code

In [3]:
# count countries
m = df.Country.value_counts()

In [5]:
pip install apyori

In [6]:
from apyori import apriori

**Warning: The below Run Time is Above 30 mins**

In [7]:
# for i in m.index:
#     new_df, new_stock_code = combine_stock_code_country(df, i)
#     print("new df for", i, "with", new_df.shape, "items")
#     print("Aprior Algorithm mine top 5 association in each country")
#     print(list(apriori(new_stock_code,
#         # min_support=0.2,
#         min_confidence=0.3
#         # min_lift=1.0,
#         # max_length=None))
# )))
#     print("end for", i)

new df for United Kingdom with (23494, 2) items
Aprior Algorithm mine top 5 association in each country
[]
end for United Kingdom
new df for Germany with (603, 2) items
Aprior Algorithm mine top 5 association in each country
[RelationRecord(items=frozenset({'POST'}), support=0.6351575456053068, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST'}), confidence=0.6351575456053068, lift=1.0)]), RelationRecord(items=frozenset({'22328', '22326'}), support=0.10281923714759536, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'22328'}), confidence=0.5210084033613446, lift=4.027795733678087), OrderedStatistic(items_base=frozenset({'22328'}), items_add=frozenset({'22326'}), confidence=0.7948717948717949, lift=4.027795733678087)]), RelationRecord(items=frozenset({'POST', '22326'}), support=0.17081260364842454, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confiden

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



KeyboardInterrupt: ignored

To prevent crash, I decide to do it each country by country. not use for loop anymall. 

In [8]:
m.index

Index(['United Kingdom', 'Germany', 'France', 'EIRE', 'Spain', 'Netherlands',
       'Belgium', 'Switzerland', 'Portugal', 'Australia', 'Norway', 'Italy',
       'Channel Islands', 'Finland', 'Cyprus', 'Sweden', 'Unspecified',
       'Austria', 'Denmark', 'Japan', 'Poland', 'Israel', 'USA', 'Hong Kong',
       'Singapore', 'Iceland', 'Canada', 'Greece', 'Malta',
       'United Arab Emirates', 'European Community', 'RSA', 'Lebanon',
       'Lithuania', 'Brazil', 'Czech Republic', 'Bahrain', 'Saudi Arabia'],
      dtype='object')

In [9]:
i = m.index[0]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

new df for United Kingdom with (23494, 2) items
Aprior Algorithm mine top 5 association in each country
[]
end for United Kingdom


In [17]:
i = m.index[1]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)



i = m.index[2]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)





i = m.index[3]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[4]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[5]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[6]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[7]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[8]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[9]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)




new df for Germany with (603, 2) items
Aprior Algorithm mine top 5 association in each country
[RelationRecord(items=frozenset({'22326'}), support=0.19734660033167495, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22326'}), confidence=0.19734660033167495, lift=1.0)]), RelationRecord(items=frozenset({'POST'}), support=0.6351575456053068, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST'}), confidence=0.6351575456053068, lift=1.0)]), RelationRecord(items=frozenset({'22328', '22326'}), support=0.10281923714759536, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'22328'}), confidence=0.5210084033613446, lift=4.027795733678087), OrderedStatistic(items_base=frozenset({'22328'}), items_add=frozenset({'22326'}), confidence=0.7948717948717949, lift=4.027795733678087)]), RelationRecord(items=frozenset({'POST', '22326'}), support=0.17081260364842454, ordered_statistics=[OrderedSt

In [20]:
i = m.index[10]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[11]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[12]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[13]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)





new df for Norway with (40, 2) items
Aprior Algorithm mine top 5 association in each country
[RelationRecord(items=frozenset({'21481'}), support=0.175, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21481'}), confidence=0.175, lift=1.0)]), RelationRecord(items=frozenset({'21731'}), support=0.15, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21731'}), confidence=0.15, lift=1.0)]), RelationRecord(items=frozenset({'21914'}), support=0.15, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21914'}), confidence=0.15, lift=1.0)]), RelationRecord(items=frozenset({'22423'}), support=0.25, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.25, lift=1.0)]), RelationRecord(items=frozenset({'22624'}), support=0.175, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22624'}), confidence=0.175, lift=1.0)]), Re

In [22]:
# takes too long time abover 30 mins


i = m.index[14]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.8
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)




new df for Cyprus with (20, 2) items
Aprior Algorithm mine top 5 association in each country


KeyboardInterrupt: ignored

In [26]:

i = m.index[15]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.1
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

new df for Sweden with (46, 2) items
Aprior Algorithm mine top 5 association in each country
[RelationRecord(items=frozenset({'21731'}), support=0.10869565217391304, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'21731'}), confidence=0.10869565217391304, lift=1.0)]), RelationRecord(items=frozenset({'22467'}), support=0.10869565217391304, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22467'}), confidence=0.10869565217391304, lift=1.0)]), RelationRecord(items=frozenset({'22492'}), support=0.13043478260869565, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22492'}), confidence=0.13043478260869565, lift=1.0)]), RelationRecord(items=frozenset({'22720'}), support=0.15217391304347827, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22720'}), confidence=0.15217391304347827, lift=1.0)]), RelationRecord(items=frozenset({'22779'}), support=0.1304347826086

In [27]:
# takes too long time abover 30 mins

i = m.index[16]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


new df for Unspecified with (13, 2) items
Aprior Algorithm mine top 5 association in each country


KeyboardInterrupt: ignored

In [29]:

i = m.index[17]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[18]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

new df for Austria with (19, 2) items
Aprior Algorithm mine top 5 association in each country


IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [ ]:

i = m.index[19]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)



i = m.index[20]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

In [ ]:
i = m.index[21]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)



i = m.index[22]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)





i = m.index[23]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[24]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[25]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[26]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[27]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[28]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[29]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[30]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

new df for Israel with (9, 2) items
Aprior Algorithm mine top 5 association in each country


In [ ]:
i = m.index[31]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)



i = m.index[32]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)





i = m.index[33]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[34]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[35]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[36]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


i = m.index[37]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)

i = m.index[38]
new_df, new_stock_code = combine_stock_code_country(df, i)
print("new df for", i, "with", new_df.shape, "items")
print("Aprior Algorithm mine top 5 association in each country")
print(list(apriori(new_stock_code,
    # min_support=0.2,
    min_confidence=0.15
    # min_lift=1.0,
    # max_length=None))
)))
print("end for", i)


### Discussion


I find top 5 association is 21731, 22326, 20725, 22423, 21481. 